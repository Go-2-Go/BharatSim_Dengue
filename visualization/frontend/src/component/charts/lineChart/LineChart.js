import React from 'react';
import PropTypes from 'prop-types';
import { Line } from 'react-chartjs-2';

import { url } from '../../../utils/url';
import useFetch from '../../../hook/useFetch';
import chartConfigStyles from './lineChartStyling';

const options = {
  maintainAspectRatio: false,
  responsive: true,
  animation: {
    duration: 0, // general animation time
  },
  hover: {
    animationDuration: 0, // duration of animations when hovering an item
  },
  responsiveAnimationDuration: 0,
  elements: {
    line: {
      tension: 0, // disables bezier curves
    },
  },
  scales: {
    yAxes: [
      {
        id: 'first-y-axis',
        type: 'linear',
        ticks: {
          sampleSize: 10,
          min: 0,
        },
      },
    ],
    xAxes: [
      {
        ticks: {
          display: false,
        },
      },
    ],
  },
};

const LineChart = ({ config }) => {
  const csvData = useFetch({
    url: url.getDataUrl(config.dataSource),
    query: { columns: [config.xAxis, config.yAxis] },
  });

  const data = {
    labels: csvData && csvData.data[config.xAxis],
    datasets: [
      {
        ...chartConfigStyles.datasets[0],
        data: csvData && csvData.data[config.yAxis],
      },
    ],
  };

  return <Line data={data} options={options} />;
};

LineChart.displayName = 'LineChart';

LineChart.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    xAxis: PropTypes.string.isRequired,
    yAxis: PropTypes.string.isRequired,
  }).isRequired,
};

export default React.memo(LineChart);
