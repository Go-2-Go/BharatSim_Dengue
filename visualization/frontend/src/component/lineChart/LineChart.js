import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { Line } from 'react-chartjs-2';

import { url } from '../../utils/url';
import useFetch from '../../hook/useFetch';
import chartConfig from './lineChartConfig';

const options = { maintainAspectRatio: false, responsive: true };

const BasicLineChart = forwardRef(({ config }, ref) => {
  const csvData = useFetch({
    url: url.getDataUrl(config.dataSource),
    query: { columns: [config.xColumn, config.yColumn] },
  });

  const data = {
    labels: csvData && csvData.data[config.xColumn],
    datasets: [
      {
        ...chartConfig.datasets[0],
        data: csvData && csvData.data[config.yColumn],
      },
    ],
  };

  return <Line ref={ref} data={data} options={options} />;
});

BasicLineChart.displayName = 'LineChart';

BasicLineChart.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    xColumn: PropTypes.string.isRequired,
    yColumn: PropTypes.string.isRequired,
  }).isRequired,
};

export default BasicLineChart;
