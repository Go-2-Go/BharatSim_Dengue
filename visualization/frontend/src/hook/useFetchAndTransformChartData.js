import useFetch from './useFetch';
import { api } from '../utils/api';
import { chartStyleConfig, getColor } from '../modules/charts/chartStyleConfig';

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export default function useFetchAndTransformChartData(config) {
  const { xAxis: xColumn, yAxis, dataSource } = config;
  const yColumns = getYaxisNames(yAxis);

  const { data: csvData, loadingState } = useFetch(api.getData, {
    params: dataSource,
    query: { columns: [xColumn, ...yColumns] },
  });

  const data = {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
  return { data, loadingState };
}
