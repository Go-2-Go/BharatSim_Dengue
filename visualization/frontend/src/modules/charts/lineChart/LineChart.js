import React from 'react';
import PropTypes from 'prop-types';
import { Line } from 'react-chartjs-2';
import { lineChartOptions } from '../chartStyleConfig';
import useFetchAndTransformChartData from '../../../hook/useFetchAndTransformChartData';
import LoaderOrError from '../../../component/loaderOrError/LoaderOrError';

function LineChart({ config }) {
  const { data, loadingState } = useFetchAndTransformChartData(config);
  return (
    <LoaderOrError loadingState={loadingState}>
      <Line data={data} options={lineChartOptions} />
    </LoaderOrError>
  );
}

LineChart.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    xAxis: PropTypes.string.isRequired,
    yAxis: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
      }),
    ).isRequired,
  }).isRequired,
};

export default React.memo(LineChart);
