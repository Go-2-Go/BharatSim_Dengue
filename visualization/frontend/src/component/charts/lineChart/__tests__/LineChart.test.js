import { render } from '@testing-library/react';
import React from 'react';

import LineChart from '../LineChart';
import useFetchAndTransformChartData from '../../../../hook/useFetchAndTransformChartData';

jest.mock('../../../../hook/useFetchAndTransformChartData');

describe('LineChart', () => {
  beforeEach(() => {
    useFetchAndTransformChartData.mockReturnValue({
      data: { labels: [1, 2], datasets: [{ data: [2, 3] }] },
      loadingState: 'SUCCESS',
    });
  });

  it('should create a line chart with single yaxis <LineChart /> component', () => {
    const { container } = render(
      <LineChart
        config={{
          dataSource: 'dataSource',
          xAxis: 'hour',
          yAxis: [{ type: 'number', name: 'exposed' }],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should show loader while fetching data <LineChart /> component', () => {
    useFetchAndTransformChartData.mockReturnValue({ data: undefined, loadingState: 'LOADING' });
    const { container } = render(
      <LineChart
        config={{
          dataSource: 'dataSource',
          xAxis: 'hour',
          yAxis: [{ type: 'number', name: 'exposed' }],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should show error after getting error while fetching <LineChart /> component', () => {
    useFetchAndTransformChartData.mockReturnValue({ data: 'error', loadingState: 'ERROR' });
    const { container } = render(
      <LineChart
        config={{
          dataSource: 'dataSource',
          xAxis: 'hour',
          yAxis: [{ type: 'number', name: 'exposed' }],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should create a line chart with multiple yaxis <LineChart /> component', () => {
    const { container } = render(
      <LineChart
        config={{
          dataSource: 'dataSource',
          xAxis: 'hour',
          yAxis: [
            { type: 'number', name: 'exposed' },
            { type: 'number', name: 'suseptible' },
          ],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
