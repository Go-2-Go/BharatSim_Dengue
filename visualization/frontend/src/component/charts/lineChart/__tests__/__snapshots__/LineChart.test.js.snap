// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart should have fetched text in <LineChart /> component 1`] = `
<div>
  <canvas
    class="chartjs-render-monitor"
    height="0"
    style="display: block;"
    width="0"
  />
</div>
`;
